thavalon-webapp/src/components/gameComponents/playerBoard.tsx

```diff
@@ -1,5 +1,177 @@
-import React from "react";
+import React, { useEffect, useState } from "react";
+import { GameSocket, InboundMessage, InboundMessageType, OutboundMessageType } from "../../utils/GameSocket";
+import { Vote, GameMessageType, GameMessage, Snapshot } from "./constants";
+import { Spinner } from "react-bootstrap";
 
+import "../../styles/gameStyles/playerBoard.scss";
+
+
+/**
+ * Props object for the PlayerCard component. Some of these aren't used yet,
+ * pending future development.
+ */
+interface PlayerCardProps {
+    name: string
+    toggleSelected: (name: string) => void,
+    me?: boolean,
+    tabbedOut?: boolean,
+    isProposing?: boolean,
+    isSelected?: boolean,
+    vote?: Vote,
+    declaredAs?: string
+}
+
+/**
+ * Message for the tabbed out indicator message.
+ */
+interface PlayerFocusChangeMessage {
+    displayName: string,
+    visibility: VisibilityState
+}
+
+/**
+ * Board containing the player list and all related functions. This is a container
+ * for all PlayerCards and has handlers for game and lobby events.
+ */
 export function PlayerBoard(): JSX.Element {
-    return <h1>Player Board</h1>;
+    // On mount, set up event handlers for the game socket and the DOM.
+    // On unmount, clean up event handlers.
+    useEffect(() => {
+        const connection = GameSocket.getInstance();
+        connection.onGameEvent.subscribe(handleMessage);
+        connection.onLobbyEvent.subscribe(handleMessage);
+        document.onvisibilitychange = () => sendPlayerVisibilitychange();
+        return () => {
+            connection.onGameEvent.unsubscribe(handleMessage);
+            document.onvisibilitychange = null;
+        }
+    }, [])
+
+    // State for maintaining the player list.
+    const [playerList, setPlayerList] = useState<string[]>([])
+    // State maintaining selected players. These players are highlighted in green.
+    const [selectedPlayers, setSelectedPlayers] = useState(new Set<string>());
+    // State for maintaining players who are tabbed out. These players have a tab indicator.
+    const [tabbedOutPlayers, setTabbedOutPlayers] = useState(new Set<string>());
+
+    /**
+     * Generic message handler for all messages from the server
+     * @param message The InboundMessage from the server.
+     */
+    function handleMessage(message: InboundMessage): void {
+        switch (message.messageType) {
+            case InboundMessageType.Snapshot:
+                const snapshot = message.data as Snapshot
+                handleGameMessage(snapshot.log[0]);
```

Comment:
This'll always be the first message of the game, not the most recent. If we're only using snapshots for reconnection though, I think this might need to process the entire log

Reply:
I'm using this to get the proposal order, since it'll always be first and works around the issue of the client not ready to receive messages when the server sends them. We'll have to use more of the log when rebuilding after a reconnect, but I haven't implemented that yet.

Reply:
Oh ok, sounds good